drivers/volume/portworx/portworx.go

```diff
@@ -744,75 +862,174 @@ func (p *portworx) createVolumeSnapshotData(
 	})
 
 	if err != nil {
-		logrus.Errorf("Error creating the VolumeSnapshotData for snap %s: %v", snapshotName, err)
-		return nil, fmt.Errorf("Failed to create the VolumeSnapshotData for snap %s", snapshotName)
+		err = fmt.Errorf("error creating the VolumeSnapshotData for snap %s due to err: %v", snapshotName, err)
+		logrus.Println(err)
+		return nil, err
 	}
 
 	return result, nil
 }
 
 func (p *portworx) findParentPVCBySnapID(snapID string) (*v1.PersistentVolumeClaim, error) {
-	pxSnaps, err := p.volDriver.Inspect([]string{snapID})
+	snapInfo, err := p.InspectVolume(snapID)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(pxSnaps) != 1 {
-		return nil, fmt.Errorf("failed to inspect Portworx snapshot: %s. Received: %d entries in response",
-			snapID, len(pxSnaps))
-	}
-
-	pxSnap := pxSnaps[0]
-
-	if len(pxSnap.Source.Parent) == 0 {
+	if len(snapInfo.ParentID) == 0 {
 		return nil, fmt.Errorf("Portworx snapshot: %s does not have parent set", snapID)
 	}
 
-	snapParents, err := p.volDriver.Inspect([]string{pxSnap.Source.Parent})
+	snapParentInfo, err := p.InspectVolume(snapInfo.ParentID)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(snapParents) != 1 {
-		return nil, fmt.Errorf("failed to inspect snapshot parent: %s. Received: %d entries in response",
-			pxSnap.Source.Parent, len(snapParents))
-	}
-
-	snapParent := snapParents[0]
-	parentPVName := snapParent.Locator.Name
-	parentPV, err := k8s.Instance().GetPersistentVolume(parentPVName)
+	parentPV, err := k8s.Instance().GetPersistentVolume(snapParentInfo.VolumeName)
 	if err != nil {
 		return nil, err
 	}
 
 	return k8s.Instance().GetPersistentVolumeClaim(parentPV.Spec.ClaimRef.Name, parentPV.Spec.ClaimRef.Namespace)
 }
 
-func (p *portworx) checkCloudSnapStatus(volID string) (string, error) {
+func (p *portworx) waitForCloudSnapCompletion(op, volID string) error {
+	return wait.ExponentialBackoff(cloudsnapBackoff, func() (bool, error) {
+		_, pendingMsg, err := p.checkCloudSnapStatus(op, volID)
+		if err != nil {
+			return true, err
+		}
+
+		if len(pendingMsg) > 0 {
+			return false, nil
+		}
+
+		return true, nil
+	})
+}
+
+func (p *portworx) checkCloudSnapStatus(op, volID string) (string /* cs ID */, string /* pending msg */, error) {
```

Comment:
Instead of the callers deciding on the pending message, return a specific/customer error that the caller should check to decide if it is pending.

Reply:
Or return a typed status, Done, Failed, Pending, etc

Reply:
updated the function